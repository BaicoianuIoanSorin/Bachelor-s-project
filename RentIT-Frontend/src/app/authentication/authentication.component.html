<div class="authentication-container">
  <nb-card size="medium" [ngStyle]="!isLoggedInTemplate ? { height: '80vh'} : {}">
    <nb-card-body>
      <ng-container *ngIf="isLoggedInTemplate; then login; else register"></ng-container>
    </nb-card-body>
  </nb-card>
</div>

<ng-template #register>
  <h1> Register </h1>
  <p> Already having an account? <a (click)="switchTemplates()">Log in</a></p>
  <div class="form-fields">
    <nb-form-field>
      <input [(ngModel)]="userContent.email" name="email" type="email" nbInput shape="round" placeholder="Email"
             (input)="onInputType($event)" [class.invalid-field]="!isEmailValid"
             nbTooltip="The email is not valid. It needs to contain '@','.'"
             [nbTooltipDisabled]="isEmailValid"
             nbTooltipPlacement="bottom"
             (nbTooltipShowStateChange)="!isEmailValid ? showTooltip() : hideTooltip()"
             #tooltipEmail="nbTooltip"
             required email>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.firstName" name="firstName" type="text" nbInput shape="round"
             placeholder="First name" (input)="onInputType($event)"
             required>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.lastName" name="lastName" type="text" nbInput shape="round"
             placeholder="Last name"
             required>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.location" name="location" type="text" nbInput shape="round" placeholder="Location"
             required>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.password" name="password" [type]="getInputType()" nbInput shape="round"
             (input)="onInputType($event)" [class.invalid-field]="!isPasswordValid"
             [nbTooltip]="'The password is not valid.\n The passwords do not match.\n It needs to contain at least 8 characters, one uppercase character, numbers and special characters.'"
             [nbTooltipDisabled]="isPasswordValid"
             nbTooltipPlacement="top"
             (nbTooltipShowStateChange)="!isPasswordValid ? showTooltip() : hideTooltip()"
             #tooltipPassword="nbTooltip"
             placeholder="Password" required>
      <button nbSuffix nbButton ghost (click)="toggleShowPassword()">
        <nb-icon [icon]="showPassword ? ICONS.EYE_OUTLINE : ICONS.EYE_OFF_OUTLINE" pack="eva"
                 [attr.aria-label]="showPassword ? 'hide password' : 'show password'">
        </nb-icon>
      </button>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.repeatPassword" name="repeatPassword" [type]="getInputType()"
             (input)="onInputType($event)" nbInput
             shape="round"
             placeholder="Repeat password" required>
      <button nbSuffix nbButton ghost (click)="toggleShowPassword()">
        <nb-icon [icon]="showPassword ? ICONS.EYE_OUTLINE : ICONS.EYE_OFF_OUTLINE" pack="eva"
                 [attr.aria-label]="showPassword ? 'hide password' : 'show password'">
        </nb-icon>
      </button>
    </nb-form-field>
    <button class="submit-button" [disabled]="isButtonDisabled()" nbButton hero
            [status]="(statusCode$ | async) === 401 ? 'warning' : 'primary'" shape="semi-round"
            (click)="onFormSubmit()" type="submit">Register
    </button>
  </div>
</ng-template>

<ng-template #login>
  <h1> Login </h1>
  <p> Don't have an account? <a (click)="switchTemplates()">Register</a></p>
  <div class="form-fields">
    <nb-form-field>
      <input [(ngModel)]="userContent.email" type="email" name="email" nbInput shape="round" placeholder="Email"
             required email ngModel>
    </nb-form-field>
    <nb-form-field>
      <input [(ngModel)]="userContent.password" [type]="getInputType()" nbInput type="password" shape="round"
             placeholder="Password" required>
      <button nbSuffix nbButton ghost (click)="toggleShowPassword()">
        <nb-icon [icon]="showPassword ? ICONS.EYE_OUTLINE : ICONS.EYE_OFF_OUTLINE"
                 pack="eva"
                 [attr.aria-label]="showPassword ? 'hide password' : 'show password'">
        </nb-icon>
      </button>
    </nb-form-field>
    <button class="submit-button" [disabled]="isButtonDisabled()" nbButton hero
            [status]="(statusCode$ | async) === 401 ? 'warning' : 'primary'" shape="semi-round"
            (click)="onFormSubmit()" type="submit">Login
    </button>
  </div>
</ng-template>
