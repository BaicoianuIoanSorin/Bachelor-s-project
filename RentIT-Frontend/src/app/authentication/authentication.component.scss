.authentication-container {
  margin-left: auto;
  margin-right: auto;
  width: 38%;

  nb-card {
    border-radius: 20px;
  }

  a {
    cursor: pointer;
  }

  .form-fields {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding-top: 20px;

    nb-form-field {
      width: 100%;
      max-width: 300px;
      margin-bottom: 23px;

      .invalid-field {
        border-color: var(--color-danger-500);
      }
    }

    .submit-button {
      width: 100%;
      max-width: 300px;
      margin-bottom: 23px;
    }
  }

  @media screen and (max-width: 768px) {
    width: 80%;
  }
}
