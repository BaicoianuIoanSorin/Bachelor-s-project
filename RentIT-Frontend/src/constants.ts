export function fileToBase64(file: File): Promise<string | ArrayBuffer | null> {
  return new Promise<string | ArrayBuffer | null>((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      if (event.target) {
        resolve(event.target.result);
      } else {
        reject("Error reading file.");
      }
    };

    reader.onerror = (event) => {
      reject("Error reading file.");
    };

    reader.readAsDataURL(file);
  });
}
