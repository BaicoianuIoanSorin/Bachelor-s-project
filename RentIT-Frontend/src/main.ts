import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';

import { AppModule } from './app/app.module';
import {environment} from "src/environment/environment.dev";
import {enableProdMode} from "@angular/core";

if(environment.production) {
  enableProdMode();
}

platformBrowserDynamic().bootstrapModule(AppModule)
  .catch(err => console.error(err));
