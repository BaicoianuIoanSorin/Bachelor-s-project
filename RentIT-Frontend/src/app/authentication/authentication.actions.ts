import {UserContent} from "src/app/authentication/constants/constants";

export class Register {
  static readonly type = '[Auth] Register';
  constructor(public user: UserContent) {
  }
}

export class Login {
  static readonly type = '[Auth] Login';
  constructor(public user: UserContent) {
  }
}

export class Logout {
  static readonly type = '[Auth] Logout';
  constructor() {
  }
}
