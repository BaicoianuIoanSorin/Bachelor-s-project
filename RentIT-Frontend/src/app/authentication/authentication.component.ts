import {Component, OnDestroy, OnInit, ViewChild} from '@angular/core';
import {
  defaultUserContent,
  isEmail,
  isPassword,
  isPhoneNumber,
  UserContent
} from "src/app/authentication/constants/constants";
import {ICONS, PRODUCTS_MENU_ITEM_URLS} from '../constants';
import {Select, Store} from "@ngxs/store";
import {Login, Register} from "src/app/authentication/authentication.actions";
import {Observable} from "rxjs";
import {AuthenticationSelector} from "src/app/authentication/authentication.selector";
import {Router} from "@angular/router";
import {UserService} from "src/api/user.service";
import {NbToastrService, NbTooltipDirective} from "@nebular/theme";

@Component({
  selector: 'app-authentication',
  templateUrl: './authentication.component.html',
  styleUrls: ['./authentication.component.scss']
})
export class AuthenticationComponent implements OnInit, OnDestroy {
  @Select(AuthenticationSelector.statusCode)
  statusCode$: Observable<number>

  userContent: UserContent = defaultUserContent();
  protected readonly ICONS = ICONS;
  isLoggedInTemplate: boolean = true; // false = register template is loaded
  showPassword = false;

  // validation
  isEmailValid: boolean = true;
  isPasswordValid: boolean = true;
  isPhoneNumberValid: boolean = true;
  @ViewChild('tooltipEmail') tooltipEmail: NbTooltipDirective;
  @ViewChild('tooltipPassword') tooltipPassword: NbTooltipDirective;
  @ViewChild('tooltipPhoneNumber') tooltipPhoneNumber: NbTooltipDirective;

  alive = true;

  constructor(private store: Store,
              private router: Router,
              private userService: UserService,
              private toastrService: NbToastrService) {
  }

  ngOnInit(): void {
    if (this.userService.isLoggedIn()) {
      this.toastrService.info(
        'You have been redirected to products page',
        'You are already logged in',
        {icon: ICONS.CHECKMARK_CIRCLE_OUTLINE}
      );
      this.router.navigate([PRODUCTS_MENU_ITEM_URLS.PRODUCTS]);
    }
  }

  getInputType() {
    if (this.showPassword) {
      return 'text';
    }
    return 'password';
  }

  toggleShowPassword() {
    this.showPassword = !this.showPassword;
  }

  switchTemplates() {
    this.isLoggedInTemplate = !this.isLoggedInTemplate;
  }

  isButtonDisabled() {
    if (this.isLoggedInTemplate) {
      return !this.userContent.password || !this.userContent.email;
    } else {
      let allFieldsFilled = !Object.values(this.userContent).some(value => !value);
      return !(allFieldsFilled && this.isEmailValid && this.isPasswordValid);
    }
  }

  onInputType(event) {
    switch (event.target.name) {
      case "email": {
        this.isEmailValid = this.userContent.email !== '' ? isEmail(this.userContent.email) : true;
        break;
      }
      case "repeatPassword":
      case "password": {
        this.isPasswordValid = (this.userContent.password === this.userContent.repeatPassword) && isPassword(this.userContent.password);
        break;
      }
      case "phoneNumber": {
        this.isPhoneNumberValid = this.userContent.phoneNumber !== '' && isPhoneNumber(this.userContent.phoneNumber);
        break;
      }
    }
  }

  showTooltip() {
    if (!this.isEmailValid) {
      this.tooltipEmail.show();
    } else if (!this.isPasswordValid) {
      this.tooltipPassword.show();
    } else if (!this.isPhoneNumberValid) {
      this.tooltipPhoneNumber.show();
    }
  }

  hideTooltip() {
    if (this.isEmailValid) {
      this.tooltipEmail.hide();
    } else if (this.isPasswordValid) {
      this.tooltipPassword.hide();
    } else if (this.isPhoneNumberValid) {
      this.tooltipPhoneNumber.hide();
    }
  }

  onFormSubmit() {
    if (this.isLoggedInTemplate) {
      this.store.dispatch(new Login(this.userContent));
    } else {
      this.store.dispatch(new Register(this.userContent)).toPromise().then(() => this.isLoggedInTemplate = true);
    }
  }

  ngOnDestroy(): void {
    this.alive = false;
  }
}
