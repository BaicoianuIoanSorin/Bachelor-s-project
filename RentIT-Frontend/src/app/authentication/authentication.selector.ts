import {Selector} from "@ngxs/store";
import { AuthenticationState, AuthenticationStateModel } from "src/app/authentication/authentication.state";

export class AuthenticationSelector {
  @Selector([AuthenticationState])
  static isFetching(state: AuthenticationStateModel) {
    return state.isFetching;
  }
}
