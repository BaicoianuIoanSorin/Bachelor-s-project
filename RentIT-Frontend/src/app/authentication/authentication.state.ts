import {Action, State, StateContext} from "@ngxs/store";
import {Injectable} from "@angular/core";
import {produce} from "immer";
import {Login, Logout, Register} from "src/app/authentication/authentication.actions";
import {UserService} from "src/api/user.service";
import {User} from "src/model/user";

export interface AuthenticationStateModel {
  isFetching: boolean;
}

export const defaultsState: AuthenticationStateModel = {
  isFetching: false,
}

@State<AuthenticationStateModel>({
  name: 'appPage',
  defaults: defaultsState,
})

@Injectable()
export class AuthenticationState {
  constructor(
    private userService: UserService
  ) {
  }

  @Action(Register)
  async register(
    {getState, setState}: StateContext<AuthenticationStateModel>,
    action: Register) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    let user = {
      ...action.user
    } as User;

    await this.userService.register(user);

    newState = produce(getState(), draft => {
      draft.isFetching = false;
    })
    setState(newState);
  }

  @Action(Login)
  async login(
    {getState, setState}: StateContext<AuthenticationStateModel>,
    action: Login) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    let user = {
      ...action.user
    } as User;

    await this.userService.login(user);

    newState = produce(getState(), draft => {
      draft.isFetching = false;
    })
    setState(newState);
  }

  @Action(Logout)
  async logout(
    {getState, setState}: StateContext<AuthenticationStateModel>) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);
    this.userService.logout();

    newState = produce(getState(), draft => {
      draft.isFetching = false;
    })
    setState(newState);
  }
}
