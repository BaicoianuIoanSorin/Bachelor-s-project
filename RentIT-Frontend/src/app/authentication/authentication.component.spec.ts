import {AuthenticationComponent} from "src/app/authentication/authentication.component";
import {createComponentFactory, Spectator} from "@ngneat/spectator";
import {RouterTestingModule} from "@angular/router/testing";
import {BrowserTestingModule} from "@angular/platform-browser/testing";
import {
  NbButtonModule,
  NbCardModule, NbFocusMonitor,
  NbFormFieldModule,
  NbIconModule,
  NbInputModule,
  NbLayoutDirectionService, NbMenuModule,
  NbOverlay,
  NbOverlayModule,
  NbOverlayService, NbSidebarModule,
  NbStatusService, NbThemeModule,
  NbToastrModule,
  NbToastrService,
  NbWindowModule,
  NbWindowService
} from "@nebular/theme";
import {PluginManager} from "@ngxs/store/src/plugin-manager";
import {NgxsModule} from "@ngxs/store";
import {AuthenticationState} from "./authentication.state";
import {HttpClientTestingModule} from "@angular/common/http/testing";
import {MockToastrService} from "../../mocks/modules/toastr.service.mock";
import {MockNbWindowService} from "../../mocks/modules/nbWindow.service.mock";
import {NbEvaIconsModule} from "@nebular/eva-icons";

describe('src > app > authentication > authentication.component.spec.ts', () => {
  let spectator: Spectator<AuthenticationComponent>;

  const createComponent = createComponentFactory({
    component: AuthenticationComponent,
    imports: [
      RouterTestingModule,
      BrowserTestingModule,
      HttpClientTestingModule,
      NbFormFieldModule,
      NbCardModule,
      NbIconModule,
      NbButtonModule,
      NbInputModule,
      NbEvaIconsModule,
      NbWindowModule.forRoot(),
      NbToastrModule.forRoot(),
      NbOverlayModule.forRoot(),
      NgxsModule.forRoot([AuthenticationState]),
      NbThemeModule.forRoot(), NbSidebarModule.forRoot(), NbMenuModule.forRoot()
    ],
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });
});
