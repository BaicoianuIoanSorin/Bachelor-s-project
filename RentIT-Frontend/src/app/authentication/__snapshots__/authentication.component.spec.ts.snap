// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src > app > authentication > authentication.component.spec.ts should match snapshot 1`] = `
<app-authentication
  ICONS={[Function Object]}
  __statusCode$__selector={[Function Observable]}
  alive={[Function Boolean]}
  isEmailValid={[Function Boolean]}
  isLoggedInTemplate={[Function Boolean]}
  isPasswordValid={[Function Boolean]}
  isPhoneNumberValid={[Function Boolean]}
  locationChooser={[Function TemplateRef2]}
  router={[Function Router]}
  showPassword="false"
  store={[Function Store]}
  toastrService={[Function NbToastrService]}
  tooltipEmail="undefined"
  tooltipPassword="undefined"
  tooltipPhoneNumber="undefined"
  userContent={[Function Object]}
  userService={[Function UserService]}
  windowService={[Function NbWindowService]}
>
  <div
    class="authentication-container"
  >
    <nb-card
      class="size-medium"
      size="medium"
    >
      <nb-card-body>
        <h1>
           Login 
        </h1>
        <p>
           Don't have an account? 
          <a
            class="register-redirect"
          >
            Register
          </a>
        </p>
        <div
          class="form-fields"
        >
          <nb-form-field
            class="nb-form-field-limited-width nb-form-field-size-medium"
          >
            <div
              class="nb-form-control-container"
            >
              <input
                class="email size-medium status-basic shape-round"
                email=""
                name="email"
                nbinput=""
                ngmodel=""
                placeholder="Email"
                required=""
                shape="round"
                type="email"
              />
            </div>
          </nb-form-field>
          <nb-form-field
            class="nb-form-field-limited-width nb-form-field-size-medium"
          >
            <div
              class="nb-form-control-container nb-form-field-control-with-suffix"
            >
              <input
                class="password size-medium status-basic shape-round"
                nbinput=""
                placeholder="Password"
                required=""
                shape="round"
                type="password"
              />
            </div>
            <div
              class="nb-form-field-addon nb-form-field-suffix-medium nb-form-field-addon-basic"
            >
              <button
                aria-disabled="false"
                class="appearance-ghost size-medium shape-rectangle icon-start icon-end status-basic"
                ghost=""
                nbbutton=""
                nbsuffix=""
                tabindex="0"
              >
                <nb-icon
                  aria-label="show password"
                  pack="eva"
                >
                  <svg
                    class="eva eva-eye-off-2-outline"
                    fill="currentColor"
                    height="100%"
                    viewBox="0 0 24 24"
                    width="100%"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      data-name="Layer 2"
                    >
                      <g
                        data-name="eye-off-2"
                      >
                        <rect
                          height="24"
                          opacity="0"
                          width="24"
                        />
                        <path
                          d="M17.81 13.39A8.93 8.93 0 0 0 21 7.62a1 1 0 1 0-2-.24 7.07 7.07 0 0 1-14 0 1 1 0 1 0-2 .24 8.93 8.93 0 0 0 3.18 5.77l-2.3 2.32a1 1 0 0 0 1.41 1.41l2.61-2.6a9.06 9.06 0 0 0 3.1.92V19a1 1 0 0 0 2 0v-3.56a9.06 9.06 0 0 0 3.1-.92l2.61 2.6a1 1 0 0 0 1.41-1.41z"
                        />
                      </g>
                    </g>
                  </svg>
                </nb-icon>
              </button>
            </div>
          </nb-form-field>
          <button
            aria-disabled="true"
            class="submit-button btn-disabled size-medium shape-semi-round appearance-hero status-primary"
            disabled=""
            hero=""
            nbbutton=""
            shape="semi-round"
            tabindex="-1"
            type="submit"
          >
            Login
          </button>
        </div>
      </nb-card-body>
    </nb-card>
  </div>
</app-authentication>
`;
